node/core/dispute-coordinator/src/cache.rs

```diff
@@ -0,0 +1,537 @@
+		OverlayCache {
+			cached_earliest_session: None,
+			cached_recent_disputes: None,
+			cached_candidate_votes: LruCache::new(capacity),
+			evicted_candidate_votes: HashMap::new(),
+			dirty: false,
+			capacity,
+		}
+	}
+
+	/// Update the earliest session.
+	pub fn update_earliest_sesion(&mut self, earliest_session: SessionIndex) {
+		let mut entry = PersistedCacheEntry::new(earliest_session);
+		entry.dirty();
+		self.cached_earliest_session = Some(entry);
+		self.dirty = true;
+	}
+
+	/// Get the earliest session from the cache.
+	pub fn get_earliest_session(&self) -> Option<&SessionIndex> {
+		if let Some(cache_entry) = &self.cached_earliest_session {
+			return Some(cache_entry.value())
+		}
+		None
+	}
+
+	/// Update the earliest session.
+	pub fn update_recent_disputes(&mut self, recent_disputes: RecentDisputes) {
+		let mut entry = PersistedCacheEntry::new(recent_disputes);
+		entry.dirty();
+		self.cached_recent_disputes = Some(entry);
+		self.dirty = true;
+	}
+
+	/// Clone the recent disputes, if any.
+	pub fn load_recent_disputes(&self) -> Option<RecentDisputes> {
+		if let Some(cache_entry) = self.cached_recent_disputes.clone() {
+			return Some(cache_entry.into())
+		}
+		None
+	}
+
+	/// Update votes for a candidate.
+	pub fn update_candidate_votes(
+		&mut self,
+		session: SessionIndex,
+		candidate_hash: &CandidateHash,
+		candidate_votes: Option<CandidateVotes>,
+	) {
+		let mut candidate_votes = PersistedCacheEntry::new(candidate_votes);
+		candidate_votes.dirty();
+
+		// Check if this is an insert and exceeds capacity.
+		if !self.cached_candidate_votes.contains_key(&(session, *candidate_hash)) {
+			if self.cached_candidate_votes.len() >= self.capacity {
+				// We need to move an entry to evicted entry list.
+				if let Some(evicted_entry) = self.cached_candidate_votes.remove_lru() {
+					if evicted_entry.1.is_dirty() {
+						// Queue the entry for writing to DB.
+						self.evicted_candidate_votes.insert(evicted_entry.0, evicted_entry.1);
+					}
+				}
+			}
+		} else {
+			println!("{:?} is known key", (session, *candidate_hash));
+		}
+
+		self.cached_candidate_votes.insert((session, *candidate_hash), candidate_votes);
+		self.dirty = true;
+	}
+
+	/// Get the candidate votes for the specific session-candidate pair, if present in cache.
+	/// The extra option wrapper in the result is to implement the `None` -> `Delete` semantics
+	/// when convering to `BackendWriteOp`.
+	///
+	/// TODO (asap) - refactor to use an enum instead of `Option<Option<CandidateVotes>>`.
+	pub fn load_candidate_votes(
+		&mut self,
+		session: SessionIndex,
+		candidate_hash: &CandidateHash,
+	) -> Option<Option<CandidateVotes>> {
+		let mut entry = self.cached_candidate_votes.remove(&(session, *candidate_hash));
+
+		if entry.is_none() {
+			entry = self.evicted_candidate_votes.remove(&(session, *candidate_hash));
+			if entry.is_none() {
+				return None
+			}
+		}
+
+		let entry = entry.expect("tested above; qed");
```

Comment:
Can't we get rid of this expect, by just using pattern matching above?

Reply:
It will never panic, but yeah, to play it safe I'll use a match.